\section{Overview}

This tutorial provides the first protocol from our recent publication \citep{Hoehna2017a}.
The second protocol is described in the \href{https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_Partition_Tutorial/RB_Partition_Tutorial.pdf}{Partitioned data analysis tutorial} and the third protocol is described in the \href{https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_BayesFactor_Tutorial/RB_BayesFactor_Tutorial.pdf}{Bayes factor tutorial}.

The present tutorial demonstrates how to set up and perform analyses using common nucleotide substitution models. 
The substitution models used in molecular evolution are continuous time Markov models, which are fully characterized by their instantaneous-rate matrix:
\begin{equation*}
Q = \begin{pmatrix} -\mu_A & \mu_{GA} & \mu_{CA} & \mu_{TA} \\
\mu_{AG} & -\mu_G  & \mu_{CG} & \mu_{TG} \\
\mu_{AC} & \mu_{GC} & -\mu_C  & \mu_{TC} \\
\mu_{AT} & \mu_{GT} & \mu_{CT} & -\mu_T 
\end{pmatrix} \mbox{  ,}
\end{equation*}
where $\mu_{ij}$ represents the instantaneous rate of substitution from state $i$ to state $j$.
The diagonal elements $\mu_i$ are the rates of \emph{not} changing out of state $i$, equal to the sum of the elements in the corresponding row.
Given the instantaneous-rate matrix, $Q$, we can compute the corresponding transition probabilities for a branch of length $t$, $P(t)$, by exponentiating the rate matrix:
\begin{equation*}
P(t) = \begin{pmatrix}          
p_{AA}(t) & p_{GA}(t) & p_{CA}(t) & p_{TA}(t) \\
p_{AG}(t) & p_{GG}(t) & p_{CG}(t) & p_{TG}(t) \\
p_{AC}(t) & p_{GC}(t) & p_{CC}(t) & p_{TC}(t) \\
p_{AT}(t) & p_{GT}(t) & p_{CT}(t) & p_{TT}(t)
\end{pmatrix} = e^{Qt} = \sum_{j=0}^\infty\frac{tQ^j}{j!} \mbox{  .}
\end{equation*}
Each specific substitution model has a uniquely defined instantaneous-rate matrix, $Q$.


In this tutorial you will perform phylogeny inference under common models of DNA sequence evolution: JC, F81, HKY85, GTR, GTR+Gamma and GTR+Gamma+I.
For all of these substitution models, you will perform an MCMC analysis to estimate phylogeny and other model parameters.
The estimated trees will be unrooted trees with independent branch-length parameters. 
We will provide comments on how to modify the tutorial if you wish to estimate rooted, clock-like trees.
All the assumptions will be covered more in detail later in this tutorial.

\subsection{Requirements}
We assume that you have read and hopefully completed the following tutorials:
\begin{itemize}
\item \href{https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_Getting_Started/RB_Getting_Started.pdf}{Getting Started with \RevBayes}
\item \href{https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_Intro_Tutorial/RB_Intro_Tutorial.pdf}{Very Basic Introduction to \Rev}
\item \href{https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_Rev_Tutorial/RB_Rev_Tutorial.pdf}{General Introduction to the \Rev syntax}
\item \href{https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_MCMC_Archery_Tutorial/RB_MCMC_Archery_Tutorial.pdf}{General Introduction to MCMC using an archery example}
\item \href{https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_MCMC_Binomial_Tutorial/RB_MCMC_Binomial_Tutorial.pdf}{General Introduction to MCMC using a coin-flipping example}
\end{itemize}
Note that the \href{https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_Intro_Tutorial/RB_Intro_Tutorial.pdf}{\Rev basics tutorial} introduces the basic syntax of \Rev but does not cover any phylogenetic models.
We tried to keep this tutorial very basic and introduce all the language concepts and theory on the way.
You may only need the \href{https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_Rev_Tutorial/RB_Rev_Tutorial.pdf}{\Rev syntax tutorial} for a more in-depth discussion of concepts in \Rev.




%%%%%%%%%%%%%%
%%   Data   %%
%%%%%%%%%%%%%%
\section{Data and files}

We provide the data file(s) which we will use in this tutorial.
You may want to use your own data instead.
In the \cl{data} folder, you will find the following files
\begin{itemize}
\item
\cl{primates\_and\_galeopterus\_cytb.nex}: Alignment of the \textit{cytochrome b} subunit from 23 primates representing 14 of the 16 families (\textit{Indriidae} and \textit{Callitrichidae} are missing). Note that there is one outgroup species included: \emph{Galeopterus variegatus}.
\end{itemize}



\section{Example: Character Evolution under the Jukes-Cantor Substitution Model}

\bigskip
\subsection{Getting Started}

The first section of this exercise involves:
(1) setting up a Jukes-Cantor (JC) substitution model for an alignment of the cytochrome b subunit;
(2) approximating the posterior probability of the tree topology and node ages (and all other parameters) using MCMC, and; 
(3) summarizing the MCMC output by computing the maximum \textit{a posteriori} tree. 

\begin{figure}[h!]
\centering
\fbox{\includegraphics[width=\textwidth,angle=0]{\ResourcePath figures/jc_graphical_model.pdf}}
\caption{\small Graphical model representation of a simple phylogenetic model. 
The graphical model shows the dependencies between the parameters.
Here, the rate matrix $Q$ is a constant variable because it is fixed and does not depend on any parameters.
The only free parameters of this model, the Jukes-Cantor model, are the tree $\Psi$ including the node ages.}
\label{fig:jc}
\end{figure}

We first consider the simplest substitution model described by \cite{Jukes1969}.
The instantaneous-rate matrix for the JC substitution model is defined as
\begin{equation*}
Q_{JC69} = \begin{pmatrix} 
{*} & \frac{1}{3} & \frac{1}{3} & \frac{1}{3} \\ 
\frac{1}{3} & {*} & \frac{1}{3} & \frac{1}{3} \\ 
\frac{1}{3} & \frac{1}{3} & {*} & \frac{1}{3} \\ 
\frac{1}{3} & \frac{1}{3} & \frac{1}{3} & {*}  
\end{pmatrix} \mbox{  ,}
\end{equation*}
which has the advantage that the transition probability matrix can be computed analytically
\begin{equation*}
P_{JC69} = \begin{pmatrix} {{1\over4} + {3\over4}e^{-rt}} & {{1\over4} - {1\over4}e^{-rt}} & {{1\over4} - {1\over4}e^{-rt}} & {{1\over4} - {1\over4}e^{-rt}} \\\\ {{1\over4} - {1\over4}e^{-rt}} & {{1\over4} + {3\over4}e^{-rt}} & {{1\over4} - {1\over4}e^{-rt}} & {{1\over4} - {1\over4}e^{-rt}} \\\\ {{1\over4} - {1\over4}e^{-rt}} & {{1\over4} - {1\over4}e^{-rt}} & {{1\over4} + {3\over4}e^{-rt}} & {{1\over4} - {1\over4}e^{-rt}} \\\\ {{1\over4} - {1\over4}e^{-rt}} & {{1\over4} - {1\over4}e^{-rt}} & {{1\over4} - {1\over4}e^{-rt}} & {{1\over4} + {3\over4}e^{-rt}}  
\end{pmatrix} \mbox{  ,}
\end{equation*}
where $t$ is the branch length in units of time, and $r$ is the rate (clock) for the process.
In the later exercises you will be asked to specify more complex substitution models.
\textbf{Don't worry, you won't have to calculate all of the transition probabilities, because
\RevBayes will take care of all the computations for you.} 
Here we only provide some of the equations for the models in case you might be interested in the details.
You will be able to complete the exercises without understanding the underlying math.


\noindent \\ \impmark The files for this example analysis are provided for you, which can easily be run using the \cl{source()} function in the \RevBayes console:
{\tt \begin{snugshade*}
\begin{lstlisting}
source("scripts/mcmc_JC.Rev")
\end{lstlisting}
\end{snugshade*}}

If everything loaded properly, then you should see the program initiate the Markov chain Monte Carlo analysis that estimates the posterior distribution. 
If you continue to let this run, then you will see it output the states of the Markov chain once the MCMC analysis begins. 

Ultimately, this is how you will execute most analyses in \RevBayes, with the full specification of the model and analyses contained in the sourced files. 
You could easily run this entire analysis on your own data by substituting your data file name for that in the model-specification file. 
However, it is important to understand the components of the model to be able to take full advantage of the flexibility and richness of \RevBayes.
Furthermore, without inspecting the \Rev scripts sourced in \cl{mcmc\_JC.Rev}, you may end up inadvertently performing inappropriate analyses on your dataset, which would be a waste of your time and CPU cycles. 
The next steps will walk you through the full specification of the model and MCMC analyses. 

\bigskip

\subsection{Loading the Data}

\noindent \\ \impmark Download data and output files (if you don't have them already) from: \\ \href{http://revbayes.github.io/tutorials.html}{http://revbayes.github.io/tutorials.html}


First load in the sequences using the \cl{readDiscreteCharacterData()} function. 
{\tt \begin{snugshade*}
\begin{lstlisting}
data <- readDiscreteCharacterData("data/primates_and_galeopterus_cytb.nex")
\end{lstlisting}
\end{snugshade*}}
Executing these lines initializes the data matrix as the respective \Rev variables. 
To report the current value of any variable, simply type the variable name and press enter. For the \cl{data} matrix, this provides information about the alignment:
{\tt \begin{snugshade*}
\begin{lstlisting}
data
|*   DNA character matrix with 23 taxa and 1141 characters
|*   =====================================================
|*   Origination:                      primates_and_galeopterus_cytb.nex
|*   Number of taxa:                   23
|*   Number of included taxa:          23
|*   Number of characters:             1141
|*   Number of included characters: 1141
|*   Datatype:                         DNA
\end{lstlisting}
\end{snugshade*}}


Next we will specify some useful variables based on our dataset. The variable \cl{data} has \emph{member functions} that we can use to retrieve information about the dataset. 
These include, for example, the number of species and the taxa.
We will need that taxon information for setting up different parts of our model.
{\tt \begin{snugshade*}
\begin{lstlisting}
n_species <- data.ntaxa()
n_branches <- 2 * n_species - 3
taxa <- data.taxa()
\end{lstlisting}
\end{snugshade*}}

%Additionally, we set up a counter variable for the number of moves that we already added to our analysis.
Additionally, we set up a vector for the moves in our analysis.
[Recall that moves are algorithms used to propose new parameter values during the MCMC simulation.]
This will make it much easier if we extend the model or analysis to include additional moves or to remove some moves.
%Similarly, we set up a counter variable for the number of monitors. 
Similarly, we set up a vector for the monitors. 
[Monitors print the values of model parameters to the screen and/or log files during the MCMC analysis].
{\tt \begin{snugshade*}
\begin{lstlisting}
moves = VectorMoves() 
monitors = VectorMonitors()
\end{lstlisting}
\end{snugshade*}}
You may have noticed that we used the \cl{=} operator to create the vector. %the move index.
This simply means that the variable is not part of the model.
You will later see that we use this operator more often, \EG when we create moves and monitors.

With the data loaded, we can now proceed to specify our Jukes-Cantor substitution model.

\subsection{Jukes-Cantor Substitution Model}

A given substitution model is defined by its corresponding instantaneous-rate matrix, $Q$.
The Jukes-Cantor substitution model does not have any free parameters (as the substitution rates are all assumed to be equal, and there is a separate parameter that scales their overall magnitude), so we can define it as a constant variable.
The function \cl{fnJC(n)} will create an instantaneous-rate matrix for a character with $n$ states.
Since we use DNA data here, we create a 4x4 instantaneous-rate matrix:
{\tt \begin{snugshade*}
\begin{lstlisting}
Q <- fnJC(4) 
\end{lstlisting}
\end{snugshade*}}
You can see the rates of the $Q$ matrix by typing
{\tt \begin{snugshade*}
\begin{lstlisting}
Q
|*   [ [ -1.0000, 0.3333, 0.3333, 0.3333 ] ,
|*     0.3333, -1.0000, 0.3333, 0.3333 ] ,
|*     0.3333, 0.3333, -1.0000, 0.3333 ] ,
|*     0.3333, 0.3333, 0.3333, -1.0000 ] ]
\end{lstlisting}
\end{snugshade*}}
As you can see, all substitution rates are equal.


\bigskip
\subsection{Tree Topology and Branch Lengths}

The tree topology and branch lengths are stochastic nodes in our phylogenetic model. 
In Figure \ref{fig:jc}, the tree topology is denoted $\Psi$ and the length of the branch leading to node $i$ is $bl_i$.

We will assume that all possible labeled, unrooted tree topologies have equal probability. 
This is the \cl{dnUniformTopology()} distribution in \RevBayes. 
Note that in \RevBayes it is advisable to specify the outgroup for your study system if you use an unrooted tree prior, whereas other software, \EG \MrBayes uses the first taxon in the data matrix file as the outgroup.
Specify the \cl{topology} stochastic node by passing in the tip labels \cl{names} to the \cl{dnUniformTopology()} distribution:
{\tt \begin{snugshade*}
\begin{lstlisting}
out_group = clade("Galeopterus_variegatus")
topology ~ dnUniformTopology(taxa, outgroup=out_group)
\end{lstlisting}
\end{snugshade*}}

Some types of stochastic nodes can be updated by a number of alternative moves. 
Different moves may explore parameter space in different ways, and it is possible to use multiple different moves for a given parameter to improve mixing (the efficiency of the MCMC simulation). 
In the case of our unrooted tree topology, for example, we can use both a nearest-neighbor interchange move (\cl{mvNNI}) and a subtree-prune and regrafting move (\cl{mvSPR}). 
These moves do not have tuning parameters associated with them, thus you only need to pass in the \cl{topology} node and proposal \cl{weight}. 
{\tt \begin{snugshade*}
\begin{lstlisting}
moves.append( mvNNI(topology, weight=1.0) )
moves.append( mvSPR(topology, weight=1.0) )
\end{lstlisting}
\end{snugshade*}}
The weight specifies how often the move will be applied either on average per iteration or relative to all other moves.
Have a look at the \href{https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_MCMC_Tutorial/RB_MCMC_Tutorial.pdf}{MCMC Diagnosis tutorial} for more details about moves and MCMC strategies (found on the \href{http://revbayes.github.io/tutorials.html}{\RevBayes~Tutorials Website}).


Next we have to create a stochastic node for each of the $2N-3$ branches in our tree (where $N=$ \cl{n\_species}). 
We can do this using a \cl{for} loop --- this is a plate in our graphical model. In this loop, we can create each of the branch-length nodes and assign each move. 
Copy this entire block of \Rev~code into the console:
{\tt \small \begin{snugshade*}
\begin{lstlisting}
for (i in 1:n_branches) {
   br_lens[i] ~ dnExponential(10.0)
   moves.append( mvScale(br_lens[i])  )
}
\end{lstlisting}
\end{snugshade*}}

It is convenient for monitoring purposes to add the tree length as deterministic variable. 
The tree length is simply the sum of all branch lengths. 
%This parameter is denoted $\mathbb{L}$ in Figure \ref{fig:jc}.
Accordingly, the tree length can be computed using the \cl{sum()} function, which calculates the sum of any vector of values.
{\tt \begin{snugshade*}
\begin{lstlisting}
TL := sum(br_lens)
\end{lstlisting}
\end{snugshade*}}

\begin{framed}
\textbf{Alternative branch-length priors}\\
Some studies, \EG \cite{Brown2010,Rannala2012}, have criticized the exponential prior distribution for branch lengths because it induces a gamma-dsitributed tree-length and the mean of this gamma distribution grows with the number of taxa.
For example, we can use instead a specific gamma prior distribution (or any other distribution defined on a positive real variable) for the tree length, and then use a Dirichlet prior distribution to break the tree length into the corresponding branch lengths \citep{Zhang2012}.
{\tt \begin{snugshade*}
\begin{lstlisting}
# specify a prior distribution on the tree length with your desired mean
TL ~ dnGamma(2,4)
moves.append( mvScale(TL)  )

# now create a random variable for the relative branch lengths
rel_branch_lengths ~ dnDirichlet( rep(1.0,n_branches) )
moves.append( mvBetaSimplex(rel_branch_lengths, weight=n_branches) )
moves.append( mvDirichletSimplex(rel_branch_lengths, weight=n_branches/10.0) )

# finally, transform the relative branch lengths into actual branch lengths
br_lens := rel_branch_lengths * TL
\end{lstlisting}
\end{snugshade*}}
\end{framed}


Finally, we can create a \emph{phylogram} (a phylogeny in which the branch lengths are proportional to the expected number of substitutions/site) by combining the tree topology and branch lengths.
We do this using the \cl{treeAssembly()} function, which applies the value of the $i^{th}$ member of the \cl{br\_lens} vector to the branch leading to the $i^{th}$ node in \cl{topology}. 
Thus, the \cl{psi} variable is a deterministic node: 

{\tt \begin{snugshade*}
\begin{lstlisting}
psi := treeAssembly(topology, br_lens)
\end{lstlisting}
\end{snugshade*}}

\begin{framed}
\textbf{Alternative Analysis\\Prior on Time-Trees: Tree Topology and Node Ages}\\
Alternatively, you may want to specify a prior on time-trees.
Here we will briefly indicate how to specify such an prior which will lead to inference of time trees.

The tree (the topology and node ages) is a stochastic node in our phylogenetic model. 
For simplicity, we will assume a uniform prior on both topologies and node ages.
The distribution in \RevBayes is \cl{dnUniformTimeTree()}. 

\impmark{Fore more information on tree priors, such as birth-death processes, please read the \href{https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_DiversificationRate_Tutorial/RB_DiversificationRate_Tutorial.pdf}{RB\_DiversificationRate\_Tutorial}.}

First, we need to specify the age of the tree:
{\tt \begin{snugshade*}
\begin{lstlisting}
root_age <- 10.0
\end{lstlisting}
\end{snugshade*}}
Here we simply assumed that the tree is 10.0 time units old. 
We could also specify a prior on the root age if we have fossil calibrations (see \href{https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_DivergenceTime_Calibration_Tutorial/RB_DivergenceTime_Calibration_Tutorial.pdf}{Divergence Time and Calibration Tutorial}).
Next, we specify the \cl{tree} stochastic variable by passing in the taxon information \cl{taxa} to the \cl{dnUniformTimeTree()} distribution:
{\tt \begin{snugshade*}
\begin{lstlisting}
psi ~ dnUniformTimeTree(rootAge=root_age, taxa=taxa)
\end{lstlisting}
\end{snugshade*}}

Some types of stochastic nodes can be updated by a number of alternative moves. 
Different moves may explore parameter space in different ways, and it is possible to use multiple different moves for a given parameter to improve mixing (the efficiency of the MCMC simulation). 
In the case of our rooted tree, for example, we can use both a nearest-neighbor interchange move without and with changing the node ages (\cl{mvNarrow} and \cl{mvNNI}) and a fixed-nodeheight subtree-prune and regrafting move (\cl{mvFNPR}) and its Metropolized-Gibbs variant (\cl{mvGPR}) \citep{Hoehna2008,Hoehna2012}. 
We also need moves that change the ages of the internal nodes, for example, \cl{mvSubtreeScale} and \cl{mvNodeTimeSlideUniform}.
These moves do not have tuning parameters associated with them, thus you only need to pass in the \cl{psi} node and proposal \cl{weight}. 
{\tt \begin{snugshade*}
\begin{lstlisting}
moves.append( mvNarrow(psi, weight=5.0) )
moves.append( mvNNI(psi, weight=1.0) )
moves.append( mvFNPR(psi, weight=3.0) )
moves.append( mvGPR(psi, weight=3.0) )
moves.append( mvSubtreeScale(psi, weight=3.0) )
moves.append( mvNodeTimeSlideUniform(psi, weight=15.0) )
\end{lstlisting}
\end{snugshade*}}
The weight specifies how often the move will be applied either on average per iteration or relative to all other moves.
Have a look at the MCMC tutorial for more details about moves and MCMC strategies: \href{http://revbayes.github.io/tutorials.html}{http://revbayes.github.io/tutorials.html}
\vspace{0.3cm}

\textbf{Molecular clock}\\
Additionally, in the case of time-calibrated trees, we need to add a molecular clock rate parameter.
For example, we know from empirical estimates that the molecular clock rate is about  0.01 (=1\%) per million years per site.
Nevertheless, we can estimate it here because we fixed the root age.
We use a uniform prior on the log-transform clock rate.
This specifies our lack of prior knowledge on the magnitude of the clock rate.
{\tt \begin{snugshade*}
\begin{lstlisting}
log_clock_rate ~ dnUniform(-6,1)
moves.append( mvSlide(log_clock_rate, weight=2.0) )
clock_rate := 10^log_clock_rate
\end{lstlisting}
\end{snugshade*}}

Instead, you could also fix the clock rate and estimate the root age.
\impmark{Fore more information on molecular clocks please read the \href{https://github.com/revbayes/revbayes_tutorial/raw/master/tutorial_TeX/RB_DivergenceTime_Tutorial/RB_DivergenceTime_Tutorial.pdf}{RB\_DivergenceTime\_Tutorial}.}

\end{framed}


\subsection{Putting it All Together}

We have fully specified all of the parameters of our phylogenetic model---the tree topology with branch lengths, and the substitution model that describes how the sequence data evolved over the tree with branch lengths.  
Collectively, these parameters comprise a distribution called the \textit{phylogenetic continuous-time Markov chain}, and we use the \cl{dnPhyloCTMC} constructor function to create this node.
This distribution requires several input arguments: 
(1) the \cl{tree} with branch lengths; 
(2) the instantaneous-rate matrix \cl{Q};
(3) the \cl{type} of character data.


Build the random variable for the character data (sequence alignment).
{\tt \begin{snugshade*}
\begin{lstlisting}
# the sequence evolution model
seq ~ dnPhyloCTMC(tree=psi, Q=Q, type="DNA")
\end{lstlisting}
\end{snugshade*}}


Once the \cl{PhyloCTMC} model has been created, we can attach our sequence data to the tip nodes in the tree.
{\tt \begin{snugshade*}
\begin{lstlisting}
seq.clamp(data)
\end{lstlisting}
\end{snugshade*}}
[Note that although we assume that our sequence data are random variables---they are realizations of our phylogenetic model---for the purposes of inference, we assume that the sequence data are ``clamped''.]
When this function is called, \RevBayes sets each of the stochastic nodes representing the tips of the tree to the corresponding nucleotide sequence in the alignment. 
This essentially tells the program that we have observed data for the sequences at the tips. 

Finally, we wrap the entire model to provide convenient access to the DAG. 
To do this, we only need to give the \cl{model()} function a single node. 
With this node, the \cl{model()} function can find all of the other nodes by following the arrows in the graphical model:
{\tt \begin{snugshade*}
\begin{lstlisting}
mymodel = model(Q)
\end{lstlisting}
\end{snugshade*}}

Now we have specified a simple phylogenetic analysis---each parameter of the model will be estimated from every site in our alignment.
If we inspect the contents of \cl{mymodel} we can review all of the nodes in the DAG:
{\tt \begin{snugshade*}
\begin{lstlisting}
mymodel
\end{lstlisting}
\end{snugshade*}}

\bigskip
\subsection{Performing an MCMC Analysis Under the Jukes-Cantor Model}

In this section, will describe how to set up the MCMC sampler and summarize the resulting posterior distribution of trees. 

\subsubsection{Specifying Monitors}

For our MCMC analysis, we need to set up a vector of \textit{monitors} to record the states of our Markov chain. 
The monitor functions are all called \cl{mn*}, where \cl{*} is the wildcard representing the monitor type.
First, we will initialize the model monitor using the \cl{mnModel} function. This creates a new monitor variable that will output the states for all model parameters when passed into a MCMC function. 
{\tt \begin{snugshade*}
\begin{lstlisting}
monitors.append( mnModel(filename="output/primates_cytb_JC.log", printgen=10) )
\end{lstlisting}
\end{snugshade*}}

The \cl{mnFile} monitor will record the states for only the parameters passed in as arguments. We use this monitor to specify the output for our sampled trees and branch lengths.

{\tt \begin{snugshade*}
\begin{lstlisting}
monitors.append( mnFile(filename="output/primates_cytb_JC.trees", printgen=10, psi) )
\end{lstlisting}
\end{snugshade*}}


Finally, create a screen monitor that will report the states of specified variables to the screen with \cl{mnScreen}:
{\tt \begin{snugshade*}
\begin{lstlisting}
monitors.append( mnScreen(printgen=1000, TL) )
\end{lstlisting}
\end{snugshade*}}
This monitor mostly helps us to see the progress of the MCMC run.

\subsubsection{Initializing and Running the MCMC Simulation}

With a fully specified model, a set of monitors, and a set of moves, we can now set up the MCMC algorithm that will sample parameter values in proportion to their posterior probability. 
The \cl{mcmc()} function will create our MCMC object:
{\tt \begin{snugshade*}
\begin{lstlisting}
mymcmc = mcmc(mymodel, monitors, moves)
\end{lstlisting}
\end{snugshade*}}
\iffalse
Notice that we also specified \cl{nruns=2} which means that \RevBayes will automatically run 2 independent MCMC runs.
You will find that the output is created in two files with extension \cl{\_run\_1} and \cl{\_run\_2} for each replicate and additionally the samples from both runs are combined into one file for more convenient post-processing.
\fi
We may wish to run the \cl{.burnin()} member function.
Recall that this function \textbf{does not} specify the number of states that we wish to discard from the MCMC analysis as burnin (i.e., the samples collected before the chain converges to the stationary distribution).  
Instead, the \cl{.burnin()} function specifies a \textit{completely separate} preliminary MCMC analysis that is used to tune the scale of the moves to improve mixing of the MCMC analysis.
{\tt \begin{snugshade*}
\begin{lstlisting}
mymcmc.burnin(generations=10000,tuningInterval=200)
\end{lstlisting}
\end{snugshade*}}


Now, run the MCMC:
{\tt \begin{snugshade*}
\begin{lstlisting}
mymcmc.run(generations=30000)
\end{lstlisting}
\end{snugshade*}}

When the analysis is complete, you will have the monitored files in your output directory.


Methods for visualizing the marginal densities of parameter values are not currently available in \RevBayes itself. 
Thus, it is important to use programs like \Tracer \citep{Rambaut2011} to evaluate mixing and non-convergence.

\noindent \\ \impmark Look at the file called \cl{output/primates\_cytb\_JC.log} in \Tracer. There you see the posterior distribution of the continuous parameters, \EG the tree length variable \cl{TL}.
\begin{figure}[htbp!]
\centering
\fbox{\includegraphics[width=0.4\textwidth]{\ResourcePath figures/primates_cytb_JC_TL_Trace.pdf}
\includegraphics[width=0.4\textwidth]{\ResourcePath figures/primates_cytb_JC_TL_Distribution.pdf}}
\caption{\small Left: Trace of tree-length samples for one MCMC run. 
The caterpillar-like look is a good sign.
You will also see that the effective sample size is comparably large, \IE much larger than 200.
Right: Posterior distribution of the tree length of the primate phylogeny under a Jukes-Cantor substitution model.}
\label{fig:jc_tree}
\end{figure}

\subsection{Exercise 1}

We are interested in the phylogenetic relationship of the Tarsiers. Therefore, we need to summarize the trees sampled from the posterior distribution.
\RevBayes can summarize the sampled trees by reading in the tree-trace file:
{\tt \begin{snugshade*}
\begin{lstlisting}
treetrace = readTreeTrace("output/primates_cytb_JC.trees", treetype="non-clock")
\end{lstlisting}
\end{snugshade*}}
The \cl{mapTree()} function will summarize the tree samples and write the maximum \textit{a posteriori} tree to file:
{\tt \begin{snugshade*}
\begin{lstlisting}
map_tree = mapTree(treetrace,"output/primates_cytb_JC_MAP.tree")
\end{lstlisting}
\end{snugshade*}}
\begin{figure}[htbp!]
\centering
\fbox{\includegraphics[width=0.8\textwidth]{\ResourcePath figures/primates_cytb_JC_tree.pdf}}
\caption{\small Maximum a posteriori estimate of the primate phylogeny under a Jukes-Cantor substitution model. 
The numbers at the nodes show the posterior probabilities for the clades.
We have rooted the tree at the outgroup \emph{Galeopterus\_variegatus}}
\label{fig:jc_tree}
\end{figure}
\noindent \\ \impmark Look at the file called \cl{output/primates\_cytb\_JC\_MAP.tree} in \texttt{FigTree}. We show it in Figure~\ref{fig:jc_tree}.

\pagebreak
\noindent \impmark Fill in the following table as you go through the tutorial.

Note, you can query the posterior probability of a clade being monophyletic using the following command:
{\tt \begin{snugshade*}
\begin{lstlisting}
Lemuroidea <- clade("Cheirogaleus_major", 
                    "Daubentonia_madagascariensis", 
                    "Lemur_catta", 
                    "Lepilemur_hubbardorum",
                    "Microcebus_murinus",
                    "Propithecus_coquereli",
                    "Varecia_variegata_variegata")
                    
treetrace.cladeProbability( Lemuroidea )
\end{lstlisting}
\end{snugshade*}}

\begin{Form}
\begin{table}[h!]
\centering
\caption{\small Posterior probabilities of primate phylogenetic relationships$^*$.}
\resizebox{\textwidth}{!}{%
\begin{tabular}{l c c c c}
\hline
\textbf{Model} & \textbf{Lemuroidea} & \textbf{Lorisoidea} & \textbf{Platyrrhini} & \textbf{Catarrhini} \\ 
\hline
\vspace{1mm}
Jukes-Cantor & \TextField[name=pp11,backgroundcolor={.85 .85 .85},color={1 0 0},height=4ex]{}  & \TextField[name=pp12,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{}  & \TextField[name=pp13,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{}  & \TextField[name=pp14,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{}\\
\hline
\vspace{1mm}
HKY85 & \TextField[name=pp21,backgroundcolor={.85 .85 .85},color={1 0 0},height=4ex]{}  & \TextField[name=pp22,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{}  & \TextField[name=pp23,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{}  & \TextField[name=pp24,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{} \\
\hline
\vspace{1mm}
F81 & \TextField[name=pp31,backgroundcolor={.85 .85 .85},color={1 0 0},height=4ex]{}  & \TextField[name=pp32,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{}  & \TextField[name=pp33,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{}  & \TextField[name=pp34,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{} \\
\hline
\vspace{1mm}
GTR & \TextField[name=pp41,backgroundcolor={.85 .85 .85},color={1 0 0},height=4ex]{}  & \TextField[name=pp42,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{}  & \TextField[name=pp43,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{}  & \TextField[name=pp44,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{} \\
\hline
\vspace{1mm}
GTR+$\Gamma$ & \TextField[name=pp51,backgroundcolor={.85 .85 .85},color={1 0 0},height=4ex]{}  & \TextField[name=pp52,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{}  & \TextField[name=pp53,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{}  & \TextField[name=pp54,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{} \\
\hline
\vspace{1mm}
GTR+$\Gamma$+I & \TextField[name=pp61,backgroundcolor={.85 .85 .85},color={1 0 0},height=4ex]{}  & \TextField[name=pp62,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{}  & \TextField[name=pp63,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{}  & \TextField[name=pp64,backgroundcolor={.85 .85 .85},color={0 0 1},height=4ex]{} \\
\hline
{\footnotesize{$^*$you can edit this table}}\\
\end{tabular}}
\label{tab:pp}
\end{table}
\end{Form}

\begin{table}[h!]
\centering
\caption{\small Primate species and famaly relationships.}
\begin{tabular}{l l l l}
\hline
\textbf{Species} & \textbf{Family} & \textbf{Parvorder} & \textbf{Suborder} \\ 
\hline
%Alouatta palliata & Atelidae & Platyrrhini (NWM) & Haplorrhini \\
Aotus trivirgatus & Aotidae & Platyrrhini (NWM) & Haplorrhini \\
Callicebus donacophilus & Pitheciidae & Platyrrhini (NWM) & Haplorrhini \\
Cebus albifrons & Cebidae & Platyrrhini (NWM) & Haplorrhini \\
Cheirogaleus major & Cheirogaleidae & Lemuroidea & Strepsirrhini \\
Chlorocebus aethiops & Cercopithecoidea & Catarrhini & Haplorrhini \\
Colobus guereza & Cercopithecoidea & Catarrhini & Haplorrhini \\
Daubentonia madagascariensis & Daubentoniidae & Lemuroidea & Strepsirrhini \\
Galago senegalensis & Galagidae & Lorisidae & Strepsirrhini \\
Hylobates lar & Hylobatidea & Catarrhini & Haplorrhini \\
Lemur catta & Lemuridae & Lemuroidea & Strepsirrhini \\
Lepilemur hubbardorum & Lepilemuridae & Lemuroidea & Strepsirrhini \\
Loris tardigradus & Lorisidae & Lorisidae & Strepsirrhini \\
Macaca mulatta & Cercopithecoidea & Catarrhini & Haplorrhini \\
Microcebus murinus & Cheirogaleidae & Lemuroidea & Strepsirrhini \\
Nycticebus coucang & Lorisidae & Lorisidae & Strepsirrhini \\
Otolemur crassicaudatus & Galagidae & Lorisidae & Strepsirrhini \\
Pan paniscus & Hominoidea & Catarrhini & Haplorrhini \\
Perodicticus potto & Lorisidae & Lorisidae & Strepsirrhini \\
Propithecus coquereli & Indriidae & Lemuroidea & Strepsirrhini \\
Saimiri sciureus & Cebidae & Platyrrhini (NWM) & Haplorrhini \\
Tarsius syrichta & Tarsiidae &  & Haplorrhini \\
Varecia variegata variegata & Lemuridae & Lemuroidea & Strepsirrhini \\
\hline
\end{tabular}
\label{tab:primates}
\end{table}




\newpage
\section{The Hasegawa-Kishino-Yano (HKY) 1985 Substitution Model}

The Jukes-Cantor model assumes that all substitution rates are equal, which also implies that the stationary frequencies of the four nucleotide bases are equal.
These assumptions are not very biologically reasonable, so we might wish to consider a more realistic substitution model that relaxes some of these assumptions.
For example, we might allow stationary frequencies, $\pi$, to be unequal, and allow rates of transition and transversion substitutions to differ, $\kappa$.
This corresponds to the substitution model proposed by \citet[][HKY]{Hasegawa1985}, which is specified with the following instantaneous-rate matrix: 
\begin{equation*}
Q_{HKY} = \begin{pmatrix} 
{\cdot} 			& {\pi_C} 	& {\kappa\pi_G} 			& {\pi_T} \\ 
{\pi_A} 		& {\cdot} 			& {\pi_C} 			& {\kappa\pi_T} \\ 
{\kappa\pi_A} 			& {\pi_C} 			& {\cdot} 			& {\pi_T} \\ 
{\pi_A} 			& {\kappa\pi_C} 			& {\pi_G} 	& {\cdot}  
\end{pmatrix} \mbox{  .}
\end{equation*}
[The diagonal ${\cdot}$ entries are equal to the negative sum of the elements in the corresponding row.] 

\noindent \\ \impmark Use the file \cl{mcmc\_JC.Rev} as a starting point for the HKY analysis.

Note that we are adding two new variables to our model.
We can define a variable \cl{pi} for the stationary frequencies that are drawn from a flat Dirichlet distribution by
{\tt \begin{snugshade*}
\begin{lstlisting}
pi_prior <- v(1,1,1,1) 
pi ~ dnDirichlet(pi_prior)
\end{lstlisting}
\end{snugshade*}}
Since \cl{pi} is a stochastic variable, we need to specify a move to propose updates to it.
A good move on variables drawn from a Dirichlet distribution is the \cl{mvBetaSimplex}.
This move randomly takes an element from the simplex, proposes a new value for it drawn from a Beta distribution, and then rescales all values of the simplex to sum to 1 again.
{\tt \begin{snugshade*}
\begin{lstlisting}
moves.append( mvBetaSimplex(pi, weight=2) )
moves.append( mvDirichletSimplex(pi, weight=1) )
\end{lstlisting}
\end{snugshade*}}

The second new variable is $\kappa$, which specifies the ratio of transition-transversion rates.
The $\kappa$ parameter must be a positive-real number and a natural choice as the prior distribution is the lognormal distribution:
{\tt \begin{snugshade*}
\begin{lstlisting}
kappa ~ dnLognormal(0.0, 1.0)
\end{lstlisting}
\end{snugshade*}}
Again, we need to specify a move for this new stochastic variable.
A simple scaling move should do the job.
{\tt \begin{snugshade*}
\begin{lstlisting}
moves.append( mvScale(kappa) )
\end{lstlisting}
\end{snugshade*}}

Finally, we need to create the HKY instantaneous-rate matrix using the \cl{fnHKY} function:
{\tt \begin{snugshade*}
\begin{lstlisting}
Q := fnHKY(kappa,pi)
\end{lstlisting}
\end{snugshade*}}
This should be all for the HKY model.

\noindent \\ \impmark Don't forget to change the output file names, otherwise your old analyses files will be overwritten.

\subsection{Exercise 2}

\begin{itemize}
\item With figure \ref{fig:jc} as your guide, draw the probabilistic graphical model of the HKY model.
\item Copy the file called  \cl{mcmc\_JC.Rev} and modify it by including the necessary parameters to specify the HKY substitution model.
\item Run an MCMC analysis to estimate the posterior distribution under the HKY substitution model.
\item Are the resulting estimates of the base frequencies equal? 
	If not, how much do they differ? 
	Are the estimated base frequencies similar to the empirical base frequencies? 
	The empirical base frequencies are the frequencies of the characters in the alignment, which can be computed with \RevBayes by \cl{data.getEmpiricalBaseFrequencies()}.
\item Is the inferred rate of transition substitutions higher than the rate of transversion substitutions? If so, by how much?
\item Like the HKY model, the Felsenstein 1981 (F81) substitution model has unequal stationary frequencies, but it assumes equal transition-transversion rates \citep{Felsenstein1981}.
	Can you set up the F81 model and run an analysis?
\item Complete the Table \ref{tab:pp} by reporting the posterior probabilities of phylogenetic relationships.
\end{itemize}






\section{The General Time-Reversible (GTR) Substitution Model}

The HKY substitution model can accommodate unequal base frequencies and different rates of transition and transversion substitutions.
Despite these extensions, the HKY model may still be too simplistic for many real datasets.
Here, we extend the HKY model to specify the General Time Reversible (GTR) substitution model \citep{Tavare1986}, which allows all six exchangeability rates to differ (Figure \ref{fig:gtr}).

The instantaneous-rate matrix for the GTR substitution model is:
\begin{equation*}
\resizebox{4in}{!}{%  
$Q_{GTR} = \begin{pmatrix}
{\cdot}	   & {r_{AC}\pi_C} & {r_{AG}\pi_G} & {r_{AT}\pi_T} \\
{r_{AC}\pi_A} & {\cdot}       & {r_{CG}\pi_G} & {r_{CT}\pi_T} \\
{r_{AC}\pi_A} & {r_{CG}\pi_C} & {\cdot}       & {r_{GT}\pi_T} \\
{r_{AC}\pi_A} & {r_{CT}\pi_C} & {r_{GT}\pi_G} & {\cdot}       \\
\end{pmatrix} \mbox{  ,} $}
\end{equation*}

where the six exchangeability parameters, $r_{ij}$, specify the relative rates of change between states $i$ and $j$.  


\begin{figure}[h!]
\centering
\fbox{\includegraphics[width=\textwidth,angle=0]{\ResourcePath figures/gtr_graphical_model.pdf}}
\caption{\small Graphical model representation of the General Time Reversible (GTR) phylogenetic model.}
\label{fig:gtr}
\end{figure}

The GTR model requires that we define and specify a prior on the six exchangeability rates, which we will describe using a flat Dirichlet distribution.
As we did previously for the Dirichlet prior on base frequencies, we first define a constant node specifying the vector of concentration-parameter values using the \cl{v()} function:
{\tt \begin{snugshade*}
\begin{lstlisting}
er_prior <- v(1,1,1,1,1,1) 
\end{lstlisting}
\end{snugshade*}}
This node defines the concentration-parameter values of the Dirichlet prior distribution on the exchangeability rates. 
Now, we can create a stochastic node for the exchangeability rates using the \cl{dnDirichlet()} function, which takes the vector of concentration-parameter values as an argument and the \cl{\rbdn} operator. 
Together, these create a stochastic node named \cl{er} ($\theta$ in Figure \ref{fig:gtr}): 
{\tt \begin{snugshade*}
\begin{lstlisting}
er ~ dnDirichlet(er_prior)
\end{lstlisting}
\end{snugshade*}}


The Dirichlet distribution assigns probability densities to a group of parameters: e.g., those that measure proportions and must sum to 1. 
Here, we have specified a six-parameter Dirichlet prior, where each value describes one of the six relative rates of the GTR model: 
(1) $A\leftrightarrows C$; (2) $A\leftrightarrows G$; (3) $A\leftrightarrows T$; (4) $C\leftrightarrows G$; (5) $C\leftrightarrows T$; (6) $G\leftrightarrows T$. 
The input parameters of a Dirichlet distribution are called shape (or concentration) parameters. 
The expectation and variance for each variable are related to the sum of the shape parameters.
The prior we specified above is a `flat' or symmetric Dirichlet distribution; all of the shape parameters are equal (1,1,1,1,1,1).
This describes a model that allows for equal rates of change between nucleotides, such that the expected rate for each is equal to $\frac{1}{6}$ (Figure \ref{dirichletFig}a).
We might also parameterize the Dirichlet distribution such that all of the shape parameters were equal to 100, which would also specify a prior with an expectation of equal exchangeability rates (Figure \ref{dirichletFig}b). 
However, by increasing the values of the shape parameters, \cl{er\_prior <- v(100,100,100,100,100,100)}, the Dirichlet distribution will more strongly favor equal exchangeability rates; ({\it i.e.}, a relatively {\em informative} prior).
Alternatively, we might consider an asymmetric Dirichlet parameterization that could reflect a strong prior belief that transition and transversion substitutions occur at different rates.
For example, we might specify the prior density \cl{er\_prior <- v(4,8,4,4,8,4)}.   
Under this model, the expected rate for transversions would be $\frac{4}{32}$ and that for transitions would be $\frac{8}{32}$, and there would be greater prior probability on sets of GTR rates that matched this configuration (Figure \ref{dirichletFig}c). 
Yet another aymmetric prior could specify that each of the six GTR rates had a different value conforming to a Dirichlet(2,4,6,8,10,12). 
This would lead to a different prior probability density for each rate parameter (Figure \ref{dirichletFig}d).
Without strong prior knowledge about the pattern of relative rates, however, we can better reflect our uncertainty by using a vague prior on the GTR rates. 
Notably, all patterns of relative rates have the same probability density under \cl{er\_prior <- v(1,1,1,1,1,1)}.
\begin{figure}[h!]
\centering
\fbox{\includegraphics[width=5in]{\ResourcePath figures/dirichlet_rates.pdf}}
\caption{\small Four different examples of Dirichlet priors on exchangeability rates.}
\label{dirichletFig}
\end{figure}


For each stochastic node in our model, we must also specify a proposal mechanism if we wish to estimate that parameter. 
The Dirichlet prior on our parameter \cl{er} creates a \href{http://en.wikipedia.org/wiki/Simplex}{\textit{simplex}} of values that sum to 1. 

{\tt\small \begin{snugshade*}
\begin{lstlisting}
moves.append( mvBetaSimplex(er, weight=3) )
moves.append( mvDirichletSimplex(er, weight=1) )
\end{lstlisting}
\end{snugshade*}}

We can use the same type of distribution as a prior on the 4 stationary frequencies ($\pi_A, \pi_C, \pi_G, \pi_T$) since these parameters also represent proportions. 
Specify a flat Dirichlet prior density on the base frequencies:
{\tt \begin{snugshade*}
\begin{lstlisting}
pi_prior <- v(1,1,1,1) 
pi ~ dnDirichlet(pi_prior)
\end{lstlisting}
\end{snugshade*}}

The node \cl{pi} represents the $\pi$ node in Figure \ref{fig:gtr}.
Now add the simplex scale move on the stationary frequencies to the moves vector:
{\tt \small \begin{snugshade*}
\begin{lstlisting}
moves.append( mvBetaSimplex(pi, weight=2) )
moves.append( mvDirichletSimplex(pi, weight=1) )
\end{lstlisting}
\end{snugshade*}}

We can finish setting up this part of the model by creating a deterministic node for the GTR instantaneous-rate matrix \cl{Q}. 
The \cl{fnGTR()} function takes a set of exchangeability rates and a set of base frequencies to compute the instantaneous-rate matrix used when calculating the likelihood of our model.
{\tt \begin{snugshade*}
\begin{lstlisting}
Q := fnGTR(er,pi)
\end{lstlisting}
\end{snugshade*}}




\subsection{Exercise 3}

\begin{itemize}
\item Use one of your previous analysis files---either the \cl{mcmc\_JC.Rev} or \cl{HKY.Rev}---to specify a GTR analysis in a new file called \cl{mcmc\_GTR.Rev}.
	Adapt the old analysis to be performed under the GTR substitution model. 
\item Run an MCMC analysis to estimate the posterior distribution.
\item Complete the table of the phylogenetic relationship of primates.
\end{itemize}






\section{The Discrete Gamma Model of Among Site Rate Variation}


Members of the GTR family of substitution models assume that rates are homogeneous across sites, an assumption that is often violated by real data.
We can accommodate variation in substitution rate among sites (ASRV) by adopting the discrete-gamma model \citep{Yang1994a}.
This model assumes that the substitution rate at each site is a random variable that is described by a discretized gamma distribution, which has two parameters: the shape parameter, $\alpha$, and the rate parameter, $\beta$. 
In order that we can interpret the branch lengths as the expected number of substitutions per site, this model assumes that the mean site rate is equal to 1.
The mean of the gamma is equal to $\alpha/\beta$, so a mean-one gamma is specified by setting the two parameters to be equal, $\alpha=\beta$.
This means that we can fully describe the gamma distribution with the single shape parameter, $\alpha$. 
The degree of among-site substitution rate variation is inversely proportional to the value of the $\alpha$-shape parameter.
As the value of the $\alpha$-shape increases, the gamma distribution increasingly resembles a normal distribution with decreasing variance, which therefore corresponds to decreasing levels of ASRV (Figure \ref{asrhGammaFig}).
By contrast, when the value of the $\alpha$-shape parameter is $< 1$, the gamma distribution assumes a concave distribution that concentrates most of the prior density on low rates, but retains some prior mass on sites with very high rates, which therefore corresponds to high levels of ASRV (Figure \ref{asrhGammaFig}).
Note that, when $\alpha = 1$, the gamma distribution collapses to an exponential distribution with a rate parameter equal to $\beta$.


\begin{figure}[h]
\centering
\fbox{\includegraphics[width=3.5in]{\ResourcePath figures/asrh_gamma.pdf}}
\caption{\small The probability density of mean-one gamma-distributed rates for different values of the $\alpha$-shape parameter.}
\label{asrhGammaFig}
\end{figure}

We typically lack prior knowledge regarding the degree of ASRV for a given alignment.
Accordingly, rather than specifying a precise value of $\alpha$, we can instead estimate the value of the $\alpha$-shape parameter from the data.
This requires that we specify a diffuse (relatively `uninformative') prior on the $\alpha$-shape parameter.
For this analysis, we will use a lognormal distribution with a mean parameter, \cl{alpha\_prior\_mean}, equal to \cl{5.0}, and standard deviation, \cl{alpha\_prior\_sd}, equal to 0.587405 (thus, 95\% of the prior density spans exactly one order of magnitude).

This approach for accommodating ASRV is another example of a hierarchical model (Figure \ref{fig:gtrg}).
That is, variation in substitution rates across sites is addressed by applying a site-specific rate multiplier to each of the $j$ sites, $r_j$.
These rate-multipliers are drawn from a discrete, mean-one gamma distribution; the shape of this prior distribution (and the corresponding degree of ASRV) is governed by the $\alpha$-shape parameter.
The $\alpha$-shape parameter, in turn, is treated as a lognormal distributed random variable.
Finally, the shape of the lognormal prior is governed by the mean and standard deviation parameters, which are set to fixed values.   

\begin{figure}[h!]
\centering
\fbox{\includegraphics[width=\textwidth,angle=0]{\ResourcePath figures/gtrg_graphical_model.pdf}}
\caption{\small Graphical model representation of the General Time Reversible (GTR) + Gamma phylogenetic model with invariable sites.}
\label{fig:gtrg}
\end{figure}

\subsection{Setting up the Gamma Model in \RevBayes}

Create a constant node called \cl{alpha\_prior\_mean} for the mean parameter and a constant node called \cl{alpha\_prior\_sd} for the standard deviation of the lognormal prior on the gamma-shape parameter (this is represented as the constant $m_\alpha$ and $sd_\alpha$ parameters in Figure \ref{fig:gtrg}):
{\tt\begin{snugshade*}
\begin{lstlisting}
alpha_prior_mean <- ln(5.0)
alpha_prior_sd <- 0.587405
\end{lstlisting}
\end{snugshade*}}

Then create a stochastic node called \cl{alpha} with a lognormal prior (this represents the stochastic node for the $\alpha$-shape parameter in Figure \ref{fig:gtrg}):
{\tt\begin{snugshade*}
\begin{lstlisting}
alpha ~ dnLognormal( alpha_prior_mean, alpha_prior_sd )
\end{lstlisting}
\end{snugshade*}}

The way the ASRV model is implemented involves discretizing the mean-one gamma distribution into a set number of rate categories, $k$. 
Thus, we can analytically marginalize over the uncertainty in the rate at each site. 
The likelihood of each site is averaged over the $k$ rate categories, where the rate multiplier is the mean (or median) of each of the discrete $k$ categories. 
To specify this, we need a deterministic node that is a vector that will hold the set of $k$ rates drawn from the gamma distribution with $k$ rate categories. 
The \cl{fnDiscretizeGamma()} function returns this deterministic node and takes three arguments: the shape and rate of the gamma distribution and the number of categories. 
Since we want to discretize a mean-one gamma distribution, we can pass in \cl{alpha} for both the shape and rate.

Initialize the \cl{gamma\_rates} deterministic node vector using the  \cl{fnDiscretizeGamma()} function with \cl{4} bins:
{\tt \begin{snugshade*}
\begin{lstlisting}
gamma_rates := fnDiscretizeGamma( alpha, alpha, 4 )
\end{lstlisting}
\end{snugshade*}}

Note that here, by convention, we set $k = 4$.
The random variable that controls the rate variation is the stochastic node \cl{alpha}. 
We will apply a simple scale move to this parameter.
{\tt \begin{snugshade*}
\begin{lstlisting}
moves.append( mvScale(alpha, weight=2.0) )
\end{lstlisting}
\end{snugshade*}}

Remember that you need to call the \cl{PhyloCTMC} constructor to include the new site-rate parameter:
{\tt \begin{snugshade*}
\begin{lstlisting}
seq ~ dnPhyloCTMC(tree=psi, Q=Q, siteRates=gamma_rates, type="DNA")
\end{lstlisting}
\end{snugshade*}}


\subsection{Exercise 4}

Modify the previous GTR analysis to specify the GTR+Gamma model. 
Run an MCMC simulation to estimate the posterior distribution.
\begin{itemize}
\item Is there an impact on the estimated phylogeny compared with the previous analyses? 
	Look at the MAP tree and the posterior probabilities of the clades.
\item Complete the table of the phylogenetic relationship of primates.
\end{itemize}



\newpage
\section{Modeling Invariable Sites}
All of the substitution models described so far assume that the sequence data are potentially variable.
That is, we assume that the sequence data are random variables; specifically, we assume that they are realizations of the specified \cl{PhyloCTMC} distribution. 
However, some sites may not be free to vary---when the substitution rate of a site is zero, it is said to be \emph{invariable}.
Invariable sites are often confused with \emph{invariant} sites---when each species exhibits the same state, it is said to be invariant.
The concepts are related but distinct.
If a site is truly invariable, it will necessarily give rise to an invariant site pattern, as such sites will always have a zero substitution rate.
However, an invariant site pattern may be achieved via multiple substitutions that happen to end in the same state for every species.

Here we describe an extension to our phylogenetic model to accommodate invariable sites.
Under the invariable-sites model \citep[][]{Hasegawa1985}, each site is invariable with probability \cl{pinvar}, and variable with probability $1-$\cl{pinvar}.

First, let's have a look at the data and see how many invariant sites we have:
{\tt \begin{snugshade*}
\begin{lstlisting}
data.getNumInvariantSites()
\end{lstlisting}
\end{snugshade*}}
There seem to be a substantial number of invariant sites.

Now let's specify the invariable-sites model in \RevBayes.
We need to specify the prior probability that a site is invariable.
A Beta distribution is a common choice for parameters representing probabilities.
{\tt \begin{snugshade*}
\begin{lstlisting}
pinvar ~ dnBeta(1,1)
\end{lstlisting}
\end{snugshade*}}
The \cl{Beta(1,1)} distribution is a flat prior distribution that specifies equal probability for all values between 0 and 1.

Then, as usual, we add a move to change this stochastic variable; we'll used a simple sliding window move.
{\tt \begin{snugshade*}
\begin{lstlisting}
moves.append( mvSlide(pinvar) )
\end{lstlisting}
\end{snugshade*}}

Finally, you need to call the \cl{PhyloCTMC} constructor to include the new\cl{pinvar} parameter:
{\tt \begin{snugshade*}
\begin{lstlisting}
seq ~ dnPhyloCTMC(tree=psi, Q=Q, siteRates=gamma_rates, pInv=pinvar, type="DNA")
\end{lstlisting}
\end{snugshade*}}

\subsection{Exercise 5}

\begin{itemize}
\item Extend the GTR model to account for invariable sites and run an analysis.
\item What is the estimated probability of invariable sites and how does it relate to the ratio of invariant sites to the total number of sites?
\item Extend the GTR+$\Gamma$ model to account for invariable sites and run an analysis.
\item What is the estimated probability of invariable sites now?
\item Complete the table of the phylogenetic relationship of primates.
\end{itemize} 


\bibliographystyle{sysbio}
\bibliography{\GlobalResourcePath refs}
